torch/fx/passes/runtime_assert.py

```diff
@@ -222,89 +306,171 @@ def match_symbol(symint, cb):
             # doesn't really have to be in the loop as it will only run once,
             # it just needs to happen right after the placeholders.
             # insert this after placeholders & added sym nodes, and before non-placeholders.
-            if node not in placeholders:
-                last_sym_node = last_placeholder
-                for _ in range(inserted_sym_nodes):
-                    last_sym_node = last_sym_node.next
-                with graph.inserting_before(last_sym_node.next):
-                    add_runtime_asserts(ras_by_symbol.pop(None, []))  # type: ignore[call-overload]
+            if node == first_non_placeholder:
+                add_runtime_asserts(ras_by_symbol.pop(None, []))  # type: ignore[call-overload]
+
+            # deduplicate asserts already present in graph
+            if node.target in (
+                torch._check,
+                torch.ops.aten._assert_scalar.default,
+            ):
+                if (
+                    node.args[0] == True  # trivial
+                    or (assert_expr := _get_sym_val(node.args[0])) in added_asserts
+                ):
+                    gm.graph.erase_node(node)
+                else:
+                    added_asserts.add(assert_expr)
+
+            # hash cons on symbolic values
+            if (
+                node.op == "call_function"
+                and (sym_expr := _get_sym_val(node)) is not None
+            ):
+                if (
+                    sym_expr in symbol_to_proxy  # example value is redundant
+                    or (
+                        _is_intermediate_tensor_shape_call(node)
+                        and not (sym_expr.free_symbols - symbol_to_proxy.keys())
+                    )  # shape call on intermediate tensor, turn into computation on input shapes
+                ):
+                    if _is_intermediate_tensor_shape_call(node):  # reify from input shapes
+                        symbol_to_proxy[sym_expr] = _sympy_interp(symbol_to_proxy, sym_expr)
+                        # won't try DCE-ing tensor compute here
+                    hash_node = symbol_to_proxy[sym_expr].node
+                    node.replace_all_uses_with(hash_node)
+                    gm.graph.erase_node(node)
+                    log.debug("CSE node %s -> %s for expr %s", node, hash_node, sym_expr)
+                else:
+                    symbol_to_proxy[sym_expr] = fx.Proxy(node)
+
+            if not any(ras for ras in ras_by_symbol.values()):
+                # if no runtime asserts are present, only do CSE,
+                # don't proceed further and add refinement calls
+                continue
```

Comment:
Not sure I understand this?

Reply:
I guess this was more to leave things in the current state, where the pass is skipped if runtime_asserts is empty, and no constrain size calls are added. Otherwise I run into dynamo export issues with test cases like this:
```
def test_sym_contains(self):
    def f(x, y):
        return x.size(0) in y

    gm, _ = torch._dynamo.export(f, aten_graph=True)(torch.ones(2), torch.ones(3))
```

And we end up with this graph, which errors with a "constraining non-symbols NYI" on the sym_constrain_range call:
```
def forward(self, L_x_ : torch.Tensor, L_y_ : torch.Tensor):
    l_x_ = L_x_
    l_y_ = L_y_
    sym_size_int = torch.ops.aten.sym_size.int(l_x_, 0);  l_x_ = None
    eq = torch.eq(l_y_, sym_size_int);  l_y_ = sym_size_int = None
    any_1 = torch.any(eq);  eq = None
    item = any_1.item();  any_1 = None
    cast_symbool_to_symint_guardless = torch.fx.experimental.symbolic_shapes.cast_symbool_to_symint_guardless(item)
    sym_constrain_range_default = torch.ops.aten.sym_constrain_range.default(cast_symbool_to_symint_guardless, min = 0, max = 1);  cast_symbool_to_symint_guardless = None
    return (item,)
```

If we switch the constrain_range call to u0 >= 0 and u0 <= 1 assertions, I see this error instead:
```
RuntimeError: Piecewise((1, Eq(u2, 1)), (0, True)) is not tracked with proxy for <torch.fx.experimental.proxy_tensor.PythonKeyTracer object at 0x7fdfbc2c9c00>

While executing %_assert_scalar_default : [num_users=0] = call_function[target=torch.ops.aten._assert_scalar.default](args = (%ge, Runtime assertion failed for expression u0 >= 0 on node 'ge'), kwargs = {})
Original traceback:
None
```

I'm not sure how to handle converted sym bools like this?

Reply:
OK, I don't fundamentally object to this, but the errors you reported here are legitimate problems that we eventually do need to fix.

As for the problem here, at a guess `cast_symbool_to_symint_guardless` isn't tracing correctly with proxies.